filesystems/drs/src/main/scala/cromwell/filesystems/drs/DrsPathBuilder.scala

```diff
@@ -1,23 +1,81 @@
 
+import cats.data.NonEmptyList
 import cloud.nio.impl.drs.DrsCloudNioFileSystemProvider
-import cromwell.core.path.{Path, PathBuilder}
-
-import scala.util.{Failure, Try}
+import com.typesafe.scalalogging.StrictLogging
+import cromwell.core.path.PathFactory.PathBuilders
+import cromwell.core.path.{Path, PathFactory, PreResolvePathBuilder}
 
+import scala.util.{Failure, Success, Try}
 
 case class DrsPathBuilder(fileSystemProvider: DrsCloudNioFileSystemProvider,
-                          requesterPaysProjectIdOption: Option[String]) extends PathBuilder {
+                          requesterPaysProjectIdOption: Option[String],
+                          preResolve: Boolean = false,
+                         ) extends PreResolvePathBuilder with StrictLogging {
 
   private val drsScheme: String = fileSystemProvider.getScheme
 
   override def name: String = "DRS"
 
-  override def build(pathAsString: String): Try[Path] = {
+  override def build(pathAsString: String, pathBuilders: PathBuilders): Try[Path] = {
     if (pathAsString.startsWith(s"$drsScheme://")) {
-      Try(DrsPath(fileSystemProvider.getCloudNioPath(pathAsString), requesterPaysProjectIdOption))
+      Try(createDrsOrOtherPath(pathAsString, pathBuilders))
     } else {
       Failure(new IllegalArgumentException(s"$pathAsString does not have a $drsScheme scheme."))
     }
   }
+
+  private def createDrsOrOtherPath(pathAsString: String, pathBuilders: PathBuilders): Path = {
+    def drsPath = DrsPath(fileSystemProvider.getCloudNioPath(pathAsString), requesterPaysProjectIdOption)
+    if (preResolve) {
+      maybeCreateOtherPath(pathAsString, pathBuilders) getOrElse drsPath
+    } else {
+      drsPath
+    }
+  }
+
+  private def maybeCreateOtherPath(pathAsString: String, pathBuilders: PathBuilders): Option[Path] = {
+
+    def logAttempt[A](description: String, attempt: => A): Option[A] = {
+      logTry(description, Try(attempt))
+    }
+
+    def logTry[A](description: String, tried: Try[A]): Option[A] = {
+      tried match {
+        case Success(result) => Option(result)
+        case Failure(exception) =>
+          logFailure(description, exception)
+      }
+    }
+
+    def logFailure(description: String, reason: Any): None.type = {
+      logger.debug(s"Unable to $description, will use a DrsPath to access '$pathAsString': $reason")
+      None
+    }
+
+    def logSuccess(path: Path): Option[Unit] = {
+      logger.debug(s"Converted '$pathAsString' to '$path'")
+      Option(())
+    }
+
+    val gcsPathOption = for {
+      // If passed an empty list of path builders, returns a None and does not run the conversion
+      nonEmptyPathBuilders <- NonEmptyList.fromList(pathBuilders)
+      // Right now, only pre-resolving GCS. In the future, could pull others like FTP, HTTP, S3, OSS, SRA, etc.
+      gsUriOption <- logAttempt(
+        "resolve the uri through Martha",
+        DrsResolver.getSimpleGsUri(pathAsString, fileSystemProvider.drsPathResolver).unsafeRunSync()
```

Comment:
😬 This is one of those areas where my scala is not great, but should we really use `unsafeRunSync` in a place like this?

Reply:
I think it's ok, but is a bit of code smell.

Various places that cross sync/async or API boundaries and deserve inspection:
- `IO.unsafeRunAsync()`
- `IO.unsafeRunSync()`
- `Await.result()`
- `throw Exception()`
- etc.

Cromwell has a long history and has picked up various bits of code depending on who and when someone implemented a section of code. For example, Cromwell contains a bit of Akka actors, Scala Future APIs, and Cats Effect IO in other places. Each is not individually bad, it just leads to inconsistencies where one needs to cross these API boundaries with "unsafe" calls.

One unsophisticated alternative is to pick some effect (`Future`, `cats.effect.IO`, etc.) and wire it \*completely\* through one's codebase. More sophisticated alternatives involve things like [Tagless Final](https://www.youtube.com/watch?v=XJ2NjqkWdck), etc.

Anyways, if you or @mcovarr have thoughts on improving this PR let me know. Otherwise, this is an N+1 case to look at if we ever refactor large swaths of the code.